lib/craft/clock_bound/client.ex

```diff
@@ -0,0 +1,217 @@
+  @clockbound_restart_grace_period 5 * 1_000_000_000
+
+  defstruct [
+    :earliest,
+    :latest,
+    :clock_status
+  ]
+
+  @type clock_status :: :unknown | :synchronized | :free_running | :invalid
+  @type t :: %__MODULE__{
+          earliest: DateTime.t(),
+          latest: DateTime.t(),
+          clock_status: clock_status()
+        }
+
+  @doc """
+  Returns the current clock time with error bounds and clock status.
+  """
+  def now(shm_path \\ @default_shm_path) do
+    with {:ok, data} <- read(shm_path),
+         real = :os.system_time(:nanosecond),
+         monotonic = monotonic_time(),
+         {:ok, {earliest, latest, clock_status}} <- compute_bound_at(real, monotonic, data),
+         {:ok, earliest} <- DateTime.from_unix(earliest, :nanosecond),
+         {:ok, latest} <- DateTime.from_unix(latest, :nanosecond) do
+      {:ok,
+       %__MODULE__{
+         earliest: earliest,
+         latest: latest,
+         clock_status: clock_status
+       }}
+    else
+      {:error, reason} -> {:error, reason}
+    end
+  end
+
+  @doc """
+  Compares two ClockBound windows and returns:
+
+    * `:lt` if a is strictly younger than b
+    * `:gt` if a is strictly older than b
+    * `:ov` if the ranges overlap and no strict ordering can be determined
+  """
+  def compare(%Client{} = cbe_1, %Client{} = cbe_2) do
+    cond do
+      DateTime.after?(cbe_1.earliest, cbe_2.latest) -> :gt
+      DateTime.before?(cbe_1.latest, cbe_2.earliest) -> :lt
+      true -> :ov
+    end
+  end
+
+  @doc """
+  Returns the current monotonic time in nanoseconds
+  For macOS, it uses the native NIF to get the monotonic time.
+  For other platforms, it uses the Erlang system_info function.
+  """
+  def monotonic_time do
+    os_monotonic_time = :erlang.system_info(:os_monotonic_time_source)
+
+    if os_monotonic_time[:clock_id] == :monotonic do
+      os_monotonic_time[:time]
+    else
+      timespec_to_nanosecond(Native.get_monotonic_time())
+    end
+  end
+
+  defp compute_bound_at(real, monotonic, clockbound_data) do
+    %{
+      as_of: as_of,
+      void_after: void_after,
+      clock_status: status,
+      max_drift: max_drift,
+      bound: bound
+    } = clockbound_data
+
+    # If max_drift is too large, it indicates that some issue in the clockbound data
+    if max_drift > 1_000_000_000 do
+      {:error, :max_drift_too_large}
+    else
+      # Validate whether the clockbound data read from shared memory is still valid at the time of the request by
+      # checking if the current monotonic time is lesser than the void_after time.
+      clock_status =
+        if status in [:synchronized, :free_running] do
+          cond do
+            # If the clockbound data has not been updated in a while, consider it trusted only up to a grace period of 5 sec.
+            as_of - monotonic < @clockbound_restart_grace_period -> status
+            monotonic < void_after -> :free_running
```

Comment:
```suggestion
            # anything beyond the grace period should be considered free_running
            monotonic < void_after -> :free_running
```
I don't really know what all this means, but maybe a comment like this to help follow the logic.